assets/helpers/paymentIntegrations/newPaymentFlow/oneOffContributions.js

```diff
@@ -0,0 +1,148 @@
+  // For brevity, unneeded fields are omitted
+  // responseCode: number | null,
+  // errorName: number | null,
+  message: string,
+|}
+
+export type CreatePayPalPaymentResponse = PaymentApiResponse<PayPalApiError, CreatePayPalPaymentSuccess>;
+
+// Data that should be posted to the payment API to create a Stripe charge.
+// https://github.com/guardian/payment-api/blob/master/src/main/scala/model/stripe/StripeChargeData.scala#L82
+// TODO: are we deprecating signed-in email?
+export type StripeChargeData = {|
+  paymentData: {
+    currency: IsoCurrency,
+    amount: number,
+    token: string,
+    email: string
+  },
+  acquisitionData: PaymentAPIAcquisitionData,
+|};
+
+
+// Data that should be posted to the payment API to get a url for the PayPal UI
+// where the user is redirected to so that they can authorize the payment.
+// https://github.com/guardian/payment-api/blob/master/src/main/scala/model/paypal/PaypalPaymentData.scala#L74
+export type CreatePaypalPaymentData = {|
+  currency: IsoCurrency,
+  amount: number,
+  // Specifies the url that PayPal should make a GET request to, should the user authorize the payment.
+  // Path of url should be /paypal/rest/return (see routes file)
+  returnURL: string,
+  // Specifies the url that PayPal should make a GET request to, should the user not authorize the payment.
+  cancelURL: string,
+|}
+
+// ----- Functions ----- //
+
+function unexpectedError(message: string): UnexpectedError {
+  return {
+    type: 'unexpectedError',
+    error: message,
+  };
+}
+
+function paymentApiEndpointWithMode(url: string) {
+  if (cookie.get('_test_username')) {
+    return addQueryParamsToURL(
+      url,
+      { mode: 'test' },
+    );
+  }
+
+  return url;
+}
+
+// Object is expected to have structure:
+// { type: "error", error: { failureReason: string } }, or
+// { type: "success", data: { currency: string, amount: number } }
+function paymentResultFromObject(json: Object): Promise<PaymentResult> {
+  if (json.error) {
+    const failureReason: CheckoutFailureReason = json.error.failureReason ? json.error.failureReason : 'unknown';
+    return Promise.resolve({ paymentStatus: 'failure', error: failureReason });
+  }
+  return Promise.resolve(PaymentSuccess);
+}
+
+// Sends a one-off payment request to the payment API and standardises the result
+// https://github.com/guardian/payment-api/blob/master/src/main/resources/routes#L17
+function postOneOffStripeExecutePaymentRequest(data: StripeChargeData): Promise<PaymentResult> {
+  return logPromise(fetchJson(
+    paymentApiEndpointWithMode(window.guardian.paymentApiStripeEndpoint),
+    postRequestOptions(data, 'omit', null),
+  ).then(paymentResultFromObject));
+}
+
+// Object is expected to have structure:
+// { type: "error", error: PayPalApiError }, or
+// { type: "success", data: PaypalPaymentSuccess }
+function createPayPalPaymentResponseFromObject(res: Object): CreatePayPalPaymentResponse {
+  if (res.data && res.data.approvalUrl) {
```

Comment:
should we also check that `res` is not undefined?

Reply:
not in this case - it's called after `resp.json()` (within `fetchJson`), so should only ever have an object as an argument, as reflected in the type. (errors in JSON parsing would go into the `catch()` of the promise, I believe)